sql-plugin/src/main/scala/com/nvidia/spark/rapids/spill/SpillFramework.scala

```diff
@@ -0,0 +1,1606 @@
+ * is closed.
+ *
+ * We hold the handle lock when we are spilling (performing IO). That means that no other consumer
+ * can access this spillable device handle while it is being spilled, including a second thread
+ * trying to spill and asking each of the handles whether they are spillable or not, as that
+ * requires the handle lock. We will relax this likely in follow on work.
+ *
+ * We never hold a store-wide coarse grain lock in the stores when we do IO.
+ */
+
+/**
+ * Common interface for all handles in the spill framework.
+ */
+trait StoreHandle extends AutoCloseable {
+  /**
+   * Approximate size of this handle, used in two scenarios:
+   * - Used from the host store to figure out how much host memory total it is tracking.
+   * - If sizeInBytes is 0, the object is tracked by the stores so it can be
+   *   removed on shutdown, or by handle.close, but 0-byte handles are not spillable.
+   */
+  val sizeInBytes: Long
+}
+
+trait SpillableHandle extends StoreHandle {
+  /**
+   * Method called to spill this handle. It can be triggered from the spill store,
+   * or directly against the handle.
+   * @return sizeInBytes if spilled, 0 for any other reason (not spillable, closed)
+   */
+  def spill: Long
+
+  /**
+   * Method used to determine whether a handle tracks an object that could be spilled
+   * @note At the level of `SpillableHandle`, the only requirement of spillability
+   *       is that the size of the handle is > 0. `sizeInBytes` is known at construction, and
+   *       is a `val`.
+   * @return true if currently spillable, false otherwise
+   */
+  private[spill] def spillable: Boolean = sizeInBytes > 0
+}
+
+/**
+ * Spillable handles that can be materialized on the device.
+ * @tparam T an auto closeable subclass. `dev` tracks an instance of this object,
+ *           on the device.
+ */
+trait DeviceSpillableHandle[T <: AutoCloseable] extends SpillableHandle {
+  private[spill] var dev: Option[T]
+
+  private[spill] override def spillable: Boolean = synchronized {
+    super.spillable && dev.isDefined
+  }
+}
+
+/**
+ * Spillable handles that can be materialized on the host.
+ * @tparam T an auto closeable subclass. `host` tracks an instance of this object,
+ *           on the host.
+ */
+trait HostSpillableHandle[T <: AutoCloseable] extends SpillableHandle {
+  private[spill] var host: Option[T]
+
+  private[spill] override def spillable: Boolean = synchronized {
+    super.spillable && host.isDefined
+  }
+}
+
+object SpillableHostBufferHandle extends Logging {
+  def apply(hmb: HostMemoryBuffer): SpillableHostBufferHandle = {
+    val handle = new SpillableHostBufferHandle(hmb.getLength, host = Some(hmb))
+    SpillFramework.stores.hostStore.track(handle)
+    // do we care if:
+    // handle didn't fit in the store as it is too large.
+    // we made memory for this so we are going to hold our noses and keep going
+    // we could spill `handle` at this point.
```

Comment:
I thought we spoke about this already. There are currently two cases.

1. We are running with no limit on the host memory, but a spill store limit
2. We have a host memory limit.

If we have a host memory limit, then the spill store is constrained by the host memory limit so we can ignore it here. we will not have allocated a HostMemoryBuffer that is too large to fit.

If we have unlimited host memory, then making a SpillableHostBufferHandle is there for code compatibility, but it should not be added to the host store at all. It should never spill. We have unlimited host memory.

We should also document this somehere.